[#8505] Run PyPy with Travis-CI

adiroiban opened the pull request:
See https://twistedmatrix.com/trac/ticket/8505

ryanhiebert commented on .travis.yml:
```diff
@@ -25,7 +38,23 @@ env:
+        virtualenv --python="$PYENV_ROOT/versions/pypy-$PYPY_VERSION/bin/python" "$HOME/virtualenvs/pypy-$PYPY_VERSION"
+        source "$HOME/virtualenvs/pypy-$PYPY_VERSION/bin/activate"
+      fi
   - pip install tox tox-travis
```
You don't technically need to install `tox` explicitly, since `tox-travis` depends on `tox`... :-)

markrwilliams commented on .travis.yml:
```diff
@@ -15,17 +26,27 @@ cache:
+        export PYPY_VERSION="5.3"
+        "$PYENV_ROOT/bin/pyenv" install --skip-existing "pypy-$PYPY_VERSION"
+        virtualenv --python="$PYENV_ROOT/versions/pypy-$PYPY_VERSION/bin/python" "$HOME/virtualenvs/pypy-$PYPY_VERSION"
+        source "$HOME/virtualenvs/pypy-$PYPY_VERSION/bin/activate"
```
@adiroiban while i investigate the coverage-related `RuntimeError`, you can use a crummy hack to work around it: increase the recursion limit.  10,000 seems to work on my laptop.
here's one way to do that:

```
echo "import sys; sys.setrecursionlimit(10000)" > $HOME/virtualenvs/pypy-$PYPY_VERSION/
/lib-python/2.7/sitecustomize.py
```

I hope I can figure out a more satisfactory solution!

adiroiban replied:
Thanks. I will give it a try... but first I will use pypy 5.1 as coverage is not documented to work with pypy 5.3 http://coverage.readthedocs.io/en/latest/

again... I have never used pypy so I have no idea about all these crazy version numbers in pypy and pypy3 :) ... and in pyenv I see 5.1 and 5.1.1  :)

markrwilliams commented:
@adiroiban it looks like travis did not to run on the last two commits -- was that intentional??

adiroiban commented:
> @adiroiban it looks like travis did not to run on the last two commits -- was that intentional??

Nope. Something is broken with Travis-CI as it was ignoring my push commits.

markrwilliams commented:
@adiroiban fair enough.  i'm happy to help if i can!

adiroiban commented:
I have other PR and travis-ci is not triggered ... so I need to investigate ...I have sent an email to Travis-CI support team

anyway we need to fix the coverage issue and this will also need to be deployed on our buildbot env

BanzaiMan commented on .travis.yml:
```diff
@@ -15,17 +26,29 @@ cache:
+        virtualenv --python="$PYENV_ROOT/versions/pypy-$PYPY_VERSION/bin/python" "$HOME/virtualenvs/pypy-$PYPY_VERSION"
+        source "$HOME/virtualenvs/pypy-$PYPY_VERSION/bin/activate"
+        echo "import sys; sys.setrecursionlimit(10000)" > $HOME/virtualenvs/pypy-$PYPY_VERSION/
+/lib-python/2.7/sitecustomize.py
```
What is this line supposed to do? This line introduces YAML syntax error, so Travis CI build will not run.

markrwilliams commented:
@adiroiban so it didn't _fix_ the issue but it did seem to let the code advance a little farther.  if you're willing to give it another shot, you can bump the recursion limit to 10000.  otherwise i'll report back after i've gotten a chance to run coverage & trial under a debug build of pypy.

adiroiban commented:
> if you're willing to give it another shot, you can bump the recursion limit to 10000. otherwise i'll report back after i've gotten a chance to run coverage & trial under a debug build of pypy.

The limit is already at 10.000 ... as I have just copy pasted from your comment :)

Maybe is worth getting in touch with coverage people to see if something special is needed for PyPy

markrwilliams commented:
@adiroiban sorry, i meant 100,000 - github ate my comment and i hastily retyped it.

i can reach out to the coverage people, but i'm pretty sure this is a pypy specific issue :( i'll update this PR when i know more.

rodrigc commented:
@markrwilliams: take a look at https://travis-ci.org/twisted/twisted/jobs/142311357

markrwilliams commented:
@rodrigc Sorry, I should have mentioned that I've filed an issue with PyPy about this: https://bitbucket.org/pypy/pypy/issues/2335/maximum-recursion-depth-exceeded-with
I'll update this PR when a solution is found.

rodrigc commented:
@markrwilliams How many Pypy patches for Twisted do you have?
Can you extract a subset of those changes which do **not** require `if PYPY`, and
can run completely on Python 2.7, and pass the coverage tests on Python 2.7?
If we can get a subset of those patches committed into Twisted, that will allow some forward progress.

Getting blocked because the coverage tool itself doesn't work under Pypy is :(

adiroiban commented:
The codecov patch protection was lifted see https://github.com/twisted-infra/braid/issues/213

so you can no merge pypy patches even if their coverage is not reported

markrwilliams commented:
Armin Rigo [just pushed a fix](https://bitbucket.org/pypy/pypy/issues/2335/maximum-recursion-depth-exceeded-with#comment-29086607) for the coverage issue.  Looks like we can revist it this after the next PyPy release!

adiroiban commented:
I have pushed a version with 100.000 recursion limit... but not luck.... so I am reverting that change.

Rather than having this code unmerged, I have pushed it for review as at least it can run pypy on travis ci... once a new pypy release is ready, in theory we should only update the version in the travis install.sh

When is the new pypy version planned to be released?

I had a quick look at the pypy website, but the release notes are done without any date ... very strange.

Similar with the bad habit that we have in Twisted, when a ticket is closed as fixed there is no note about when it will be released ... or in which version it was released

glyph commented:
@adiroiban - can you update your branch please?  hopefully the push of the merge commit will also fix whatever broke the documentation buildbot…

adiroiban commented:
@glyph updated. thanks!

glyph commented:
Updating again in the hopes that the CI gods will be with us this time…

glyph commented:
Ugh. They were, but now conflicts need to be resolved.

glyph commented:
@adiroiban - the build is failing because the topfile is in the wrong place. Move it into `src/` and I think we can merge this…

glyph commented on .travis/install.sh:
```diff
@@ -1,20 +1,54 @@
 fi
 
 # Temporary workaround for https://github.com/pypa/setuptools/issues/776;
-# install (and thereby cache a built wheel of) cryptography.  (NB: We're
```
Why delete this comment?  Seems irrelevant to the change in question.

adiroiban replied:
I have removed it as for pypy this is no longer valid.

In travis we have pypy 2.5 and in venv we have pypy 5.4

glyph commented on .travis/run.sh:
```diff
@@ -2,8 +2,10 @@
 set -x
 
-if [[ "$(uname -s)" == "Darwin" ]]; then
-    # Initialize the virtualenv created at install time.
```
This change also seems spurious…

adiroiban replied:
For pypy we also need to source the venv...

so instead of adding a condition for the OS / Travis ENV, I went for sourcing anything which was created at the install time.

In this way, if later we add a PYPY3 for example in install.sh we don't need to add a new source condition here in run.sh

glyph commented on .travis.yml:
```diff
@@ -19,7 +19,22 @@ branches:
+  apt:
+    packages:
+    - libssl-dev
+    - libssl1.0.0
```
Obviously this is fine because the build works, but… why don't we need `libffi-dev`?

adiroiban replied:
I don't know... but this is what we have in trunk.

glyph commented on .travis.yml:
```diff
@@ -19,7 +19,22 @@ branches:
+cache:
+  directories:
+    - $HOME/.cache/pip
+    - $HOME/.pyenv
```
Just curious - can this not be expressed as just `$PYENV_ROOT`?

adiroiban replied:
it can. done.

adiroiban commented:
Thanks @glyph for the review. I have moved to src and also updated the travis cache functionality 

It would be nice to have a command in pip which can remove packages which are not used for a long time.

for pyenv, I went for this simple pyenv uninstall command, but we might need a .travis/pyenv.cleanup.sh script which will remove any version, other than PYPY_VERSION ... and in the future we might want to keep other versions if pyenv

adiroiban commented:
we are caching on travis the full ~/.cache/pip , but maybe we should only cache ~/.cache/pip/wheels

njsmith commented:
Instead of doing all this work with pyenv and caching, wouldn't it be easier to install squeaky's portable pypy from scratch on each build? (The travis cache has to be downloaded to the build worker every time anyway, so it's only really useful for artifacts that you're building on travis.)

glyph commented:
@njsmith - An alternate PR would be appreciated, but given that "all this work" is already done, is there much benefit to getting rid of it?

njsmith commented:
I don't know. I just happened to glance at this out of curiosity and on a
quick skim it looked like there was some struggle with getting the cache
state management to work correctly -- the advantage of my suggestion is
that it'd be stateless.

markrwilliams commented:
@adiroiban I agree with @njsmith and would prefer to use [squeaky's portable PyPy](https://github.com/squeaky-pl/portable-pypy).  I can take up this PR if that approach seems OK to you.

adiroiban commented:
@markrwilliams go for it. Thanks!

I have never used pypy so my work from here was just trying to pay it forward for all the good things I use in Twisted basically great python 2.7 as I don't care for the rest :)

I think that I have just checked what Travis-CI was doing and I went on those lines.

ulidtko commented:
*/me zaps a blessed wand of dead turning on the "Twisted on PyPy" thread. The PR rises from the dead.*

glyph commented:
@ulidtko You have successfully nerd-sniped me.

glyph committed:
make pypy/pypy3 work in travis configuration

glyph committed:
various lint failures

glyph committed:
remove pypy3 from allowed failures now that it doesn't fail

glyph committed:
don't break on windows

Julian commented:
What's wrong with the travis-provided PyPy that means pyenv needs to be involved?

glyph commented:
Historically it has been unspeakably ancient, and only the most recent pypys are really worth testing. If travis is on the latest we could revert that part.

glyph commented:
(I’d rather just land this as-is and do that as a follow-up though)

Julian commented:
Got it -- yup sounds good (me personally I've [not had any issues recently
with it being terribly out of date](
https://travis-ci.org/Julian/jsonschema/jobs/569829619#L273), but yeah
think doing it as a follow up is fine)

On Fri, Aug 9, 2019 at 8:10 PM Glyph <notifications@github.com> wrote:

> (I’d rather just land this as-is and do that as a follow-up though)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/twisted/twisted/pull/226?email_source=notifications&email_token=AACQQXU343MJ3HVQOGQFVA3QDYBPZA5CNFSM4CH2RFWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4ABSUA#issuecomment-520100176>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACQQXWRAASMCHULIW5LX6LQDYBPZANCNFSM4CH2RFWA>
> .
>

Julian commented on src/twisted/python/compat.py:
```diff
@@ -852,6 +852,37 @@ def _get_async_param(isAsync=None, **kwargs):
+    realFromFD = socket.fromfd
+    if realFromFD.__name__ == fromFDWithoutModifyingFlags.__name__:
+        return
+    socket.fromfd = fromFDWithoutModifyingFlags
```
So... personally I have experienced cases where twisted has monkey patched globally a thing and that monkey patch *introduced* bugs in non-twisted code...

I realize how tempting this is over instead just changing code to use a level of indirection, since I presume fromfd is reasonably widely used but... severe :(

Tell me to go away though I guess, and that this is the best way.

glyph replied:
This is the best way, at least in this specific case.

- It's only monkeypatched on pypy3
- The behavior on pypy3 is buggy as heck and this will fix it for other programs as well.
- The monkeypatch is harmless (even if the monkeypatch were to get invoked on a different cpython, it would at worst be slightly inefficient unless someone is doing something *truly* batty with threads).
- I expect this to be relatively short-lived; it links to the bug, the bug is super straightforward and has a reproducer that's only a half a dozen lines.  The other bug I filed today, Armin has already fixed.

Julian reviewed (approved):
I really dislike the monkey patching, but other than that, lgtm.
